jerry-core/parser/js/js-parser.c

```diff
@@ -2689,42 +2784,167 @@ parser_raise_error (parser_context_t *context_p, /**< context */
+  context_p->debugger_source_info_size = (uint16_t) new_source_info_size;
+
+  *buffer_p++ = full_type;
+} /* parser_append_source_info_type */
+
+/**
+ * Append a source info with value.
+ */
+void
+parser_append_source_info_value (parser_context_t *context_p, /**< context */
+                                 uint8_t full_type, /**< type of source info with flags */
+                                 uint32_t value) /**< source value */
+{
+  JERRY_ASSERT (JERRY_CONTEXT (debugger_flags) & JERRY_DEBUGGER_CONNECTED);
+
+  if ((full_type & JERRY_DEBUGGER_SRC_TYPE_MASK) == JERRY_DEBUGGER_SRC_STATEMENT_BREAKPOINT)
   {
-    parser_send_breakpoints (context_p, type);
+    context_p->status_flags |= PARSER_DEBUGGER_BREAKPOINT_APPENDED;
+  }
+
+  uint32_t value_size = 1;
+
+  if (unlikely (value >= 0x10000))
+  {
+    value_size = 3;
+
+    if (value >= 0x1000000)
+    {
+      value_size = 4;
+    }
+  }
+  else if (value >= 0x100)
+  {
+    value_size = 2;
+  }
+
+  uint32_t new_source_info_size = context_p->debugger_source_info_size + 1u + value_size;
+
+  if (new_source_info_size > JERRY_CONTEXT (debugger_max_send_size))
+  {
+    parser_send_source_info (context_p);
+    JERRY_ASSERT (context_p->debugger_source_info_size == 0);
+
+    new_source_info_size = 1u + value_size;
+  }
+
+  uint8_t *buffer_p = context_p->debugger_source_info + context_p->debugger_source_info_size;
+
+  context_p->debugger_source_info_size = (uint16_t) new_source_info_size;
+
+  *buffer_p++ = (uint8_t) (full_type | (value_size << JERRY_DEBUGGER_SRC_LENGTH_SHIFT));
+
+  do
+  {
+    *buffer_p++ = (uint8_t) value;
+
+    value >>= 8;
+    value_size--;
+  }
+  while (value_size > 0);
+} /* parser_append_source_info_value */
+
+/**
+ * Append source location.
+ */
+void
+parser_append_source_location (parser_context_t *context_p, /**< context */
+                               uint8_t type, /**< type of source info */
+                               parser_line_counter_t line, /**< source location line */
+                               parser_line_counter_t column) /**< source location column */
+{
+  JERRY_ASSERT (type >= JERRY_DEBUGGER_SRC_STATEMENT && type <= JERRY_DEBUGGER_SRC_EXPR_END);
+
+  if (context_p->debugger_current_line != line)
+  {
+    parser_append_source_info_value (context_p, JERRY_DEBUGGER_SRC_LINE, line);
+    context_p->debugger_current_line = line;
+
+    if (!(JERRY_CONTEXT (debugger_flags) & JERRY_DEBUGGER_CONNECTED))
+    {
+      return;
+    }
+  }
+
+  if (type == JERRY_DEBUGGER_SRC_EXPR_END)
+  {
+    /* This ensures that the parser_append_source_info_type case runs below. */
+    column = context_p->debugger_current_column;
+  }
+
+  uint32_t byte_code_size = context_p->byte_code_size;
+
+  if (type == JERRY_DEBUGGER_SRC_STATEMENT_BREAKPOINT)
+  {
+    JERRY_ASSERT (context_p->debugger_current_breakpoint_line != line);
+
+    parser_emit_cbc (context_p, CBC_BREAKPOINT_DISABLED);
+    parser_flush_cbc (context_p);
```

Comment:
The new names of the helper functions don't indicate any more that breakpoint opcodes are being emitted.
Is your idea to move away from these opcodes and move towards a separate breakpoint list? (I'd be in favor of that because one downside of the current approach is that the bytecode cannot be in ROM when debugging.)